src/modelfree/shaping_wrappers.py

```diff
@@ -0,0 +1,208 @@
+
+    def _log_sparse_dense_rewards(self):
+        if self.counter == 2048 * self.num_envs:
+            num_episodes = len(self.ep_rew_dict['reward_remaining'])
+            dense_terms = ['reward_center', 'reward_ctrl', 'reward_contact', 'reward_survive']
+            for term in dense_terms:
+                assert len(self.ep_rew_dict[term]) == num_episodes
+
+            ep_dense_mean = sum([sum(self.ep_rew_dict[t]) for t in dense_terms]) / num_episodes
+            logger.logkv('epdensemean', ep_dense_mean)
+
+            ep_sparse_mean = sum(self.ep_rew_dict['reward_remaining']) / num_episodes
+            logger.logkv('epsparsemean', ep_sparse_mean)
+
+            c = self.reward_annealer()
+            ep_rew_mean = c * ep_dense_mean + (1 - c) * ep_sparse_mean
+            logger.logkv('eprewmean_true', ep_rew_mean)
+
+            logger.logkv('rew_anneal', c)
+
+            for rew_type in self.ep_rew_dict:
+                self.ep_rew_dict[rew_type] = []
+            self.counter = 0
+
+    def reset(self):
+        return self.venv.reset()
+
+    def step_wait(self):
+        obs, rew, done, infos = self.venv.step_wait()
+
+        # replace rew with differently shaped rew
+        # victim is agent 0, attacker is agent 1
+        for env_num in range(self.num_envs):
+            shaped_reward = 0
+            for rew_type, rew_value in infos[env_num][1].items():
+                if rew_type not in self.shaping_params:
+                    continue
+
+                weighted_reward = self.shaping_params[rew_type] * rew_value
+                self.step_rew_dict[rew_type][env_num].append(weighted_reward)
+                if self.reward_annealer is not None:
+                    c = self.get_annealed_exploration_reward()
+                    if rew_type == 'reward_remaining':
+                        weighted_reward *= (1 - c)
+                    else:
+                        weighted_reward *= c
+                shaped_reward += weighted_reward
+
+            if done[env_num]:
+                for rew_type in self.step_rew_dict:
+                    self.ep_rew_dict[rew_type].append(sum(self.step_rew_dict[rew_type][env_num]))
+                    self.step_rew_dict[rew_type][env_num] = []
+            self.counter += 1
+            self._log_sparse_dense_rewards()
+            rew[env_num] = shaped_reward
+
+        return obs, rew, done, infos
+
+    def get_annealed_exploration_reward(self):
+        """
+        Returns c (which will be annealed to zero) s.t. the total reward equals
+        c * reward_move + (1 - c) * reward_remaining
+        """
+        c = self.reward_annealer()
+        assert 0 <= c <= 1
+        return c
+
+
+class NoisyAgentWrapper(BaseRLModel):
+    def __init__(self, agent, noise_annealer, noise_type='gaussian'):
+        """
+        Wrap an agent and add noise to its actions
+        :param agent: PolicyToModel(BaseRLModel) the agent to wrap
```

Comment:
Do you depend on it being `PolicyToModel`? If not, the type is just `BaseRLModel`